lisa/core/node.py

```diff
@@ -130,7 +148,8 @@ def getTool(self, tool_type: Type[T]) -> T:
                     )
             else:
                 log.debug(f"{tool_prefix} is installed already")
-        return tool
+            self.tools[cast_tool_type] = tool
```

Comment:
What is this state we're recording, and why?

Reply:
The tool state saves in node, it across test suites. Remote operation is time costing, I don't want to run a couple of "command -v" when calling getTool. The gain of caching much more than moving from paramiko to ssh2-python.

Reply:
Hm, it should not cost much to re-run these commands. The highest cost is the connection, so we need to be setting up SSH ControlMaster and sharing one session for the duration of the tests, at least as much as possible.

Reply:
Oof, Paramiko doesn't support it yet (and it doesn't look like it will any time soon): https://github.com/paramiko/paramiko/issues/852

Reply:
If there is an easy cache, I don't see reason not to use it.

Below is my runs from beijing to westus, and removes some irrelative lines. It should have good bandwitdh and bad latency.

ssh connection: ~2sec
commands: 
uname: 0.509 sec
echo: 0.436 sec
uname: 0.329 sec
echo: 0.475 sec
script(including upload): 3.134 sec